middleware/request_id.go

```diff
@@ -0,0 +1,36 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package middleware
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/heketi/heketi/pkg/utils"
+)
+
+type contextKey string
+
+var requestIDKey = contextKey("X-Request-ID")
+
+type RequestID struct {
+}
+
+// GetRequestID returns the request id from HTTP context.
+func GetRequestID(ctx context.Context) string {
+	reqID, _ := ctx.Value(requestIDKey).(string)
+	return reqID
+}
+
+func (reqID *RequestID) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+
+	newCtx := context.WithValue(r.Context(), requestIDKey, utils.GenUUID())
+	next(w, r.WithContext(newCtx))
```

Comment:
This seems pretty cool. The only thing that I wonder about is if we should avoid generating these IDs for GETs. GETs should not need and IDs and they're probably the most common request. Skipping the ID assignment on a GET might be more efficient and might help us avoid accidentally making changes in response to GETs.

Reply:
Good one, Thanks! Done.